### DISCUSSION TITLE
ERROR: Create an Entity Ruler

### DISCUSSION THREAD

**Discussion by judmesa** [explosion/spaCy#10082] [Help: Coding & Implementations] [answered]
Hi, I would like to ask for your help because now I am trying to create and entity ruler but I am having the same error

-------------Code--------------
import os
import spacy
from pptx import Presentation
from spacy.matcher import Matcher
from spacy.matcher import PhraseMatcher 
from spacy.lang.en import English
from spacy.pipeline import EntityRuler
import json

nlp=spacy.load("en_core_web_md")
#Inicializa el matcher con el vocabulario compartido
phrase_matcher = PhraseMatcher(nlp.vocab)
.
.
.
def generate_rules(patterns):
    nlp=English()
    ruler=EntityRuler(nlp)
    ruler.add_patterns(patterns)
    nlp.add_pipe('ruler')
    nlp.to_disk("Stations_NER")

Error:

ValueError: [E002] Can't find factory for 'ruler' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, doc_cleaner, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, spancat, textcat_multilabel, en.lemmatizer

**Comment by pmbaumgartner** [comment, marked as answer]
Could you try adding the component with the name `"entity_ruler"` and not `"ruler"`, like this:

```
nlp.add_pipe("entity_ruler")
```

There's also a self-contained example of the entity ruler component in the docs [here](https://spacy.io/usage/rule-based-matching#entityruler-usage) you can use for reference.

If that doesn't work, it looks to me like you've excluded some of the code you might be using. It's not clear to me how `generate_rules` should be working without seeing how you're calling the function and what might exist prior to that call. Could you include a more complete snippet of your code to help us figure out what the issue might be?

**Comment by darrkj** [comment]
It looks like you are using both methods to construct an entity ruler. The first is `ruler=EntityRuler(nlp)` and the second is `nlp.add_pipe('ruler')`.  As stated in the other response the add_pipe method should receive the factory name for the component and not a local variable name, thus entity_ruler instead of ruler.

From that snippet you have I think this would resolve the error and also remove one of the two creation methods. 

```python
nlp=English()
ruler=nlp.add_pipe('entity_ruler')
ruler.add_patterns(patterns)
```

I made this notebook so that I can refer back to some of the nuance around the two methods that can create an entity_ruler if its helpful.

[ent_ruler_test.pdf](https://github.com/explosion/spaCy/files/7913633/ent_ruler_test.pdf)
